server/src/main/java/org/elasticsearch/action/admin/cluster/stats/TransportClusterStatsAction.java

```diff
@@ -315,4 +377,119 @@ protected boolean isFresh(Long currentKey, Long newKey) {
             return newKey <= currentKey;
         }
     }
+
+    private static boolean doRemotes(ClusterStatsRequest request) {
+        return CCS_TELEMETRY_FEATURE_FLAG.isEnabled() && request.doRemotes();
+    }
+
+    private class RemoteStatsFanout extends CancellableFanOut<String, RemoteClusterStatsResponse, Map<String, RemoteClusterStats>> {
+        private final ClusterStatsRequest request;
+        private final Map<String, RemoteClusterStatsResponse> responses = new ConcurrentHashMap<>();
+        private final Executor requestExecutor;
+        private final Task task;
+        private final TaskId taskId;
+
+        RemoteStatsFanout(Task task, ClusterStatsRequest request, Executor requestExecutor) {
+            this.task = task;
+            this.request = request;
+            this.requestExecutor = requestExecutor;
+            if (task instanceof CancellableTask cancellableTask) {
+                cancellableTask.addListener(responses::clear);
```

Comment:
This shouldn't be necessary, we should instead drop all refs to the `RemoteStatsFanout` instance itself on cancellation. That's kind of the point of a `CancellableFanOut`.

Reply:
My code doesn't retain any refs to the fanout class, so that should happen as soon as `run()` exits I assume?

Reply:
Technically no, there will be refs retained by the various listeners that are waiting for transport responses too which won't be released until those listeners are completed. But they do the right thing on cancellation. It's also easy to write a lambda which accidentally captures a ref too, although I don't think that applies as things are written today.